terraform-providers: Terraform 0.13 compatibility

timstott opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

Terraform 0.13 introduces a [new filesystem layout](https://www.terraform.io/upgrade-guides/0-13.html#new-filesystem-layout-for-local-copies-of-providers) to store providers which is incompatible with Terraform 0.12. The new filesystem layout is informed by the Terraform Registry, a repository of providers and modules.

The objectives of this work are:
- Make existing terraform providers available to Terraform 0.13 where possible
- Change the update script to use the Terraform Registry

Modified
- Added a `provider-source-address` attribute to providers to enable the creation of the filesystem layout.
- The update provider script
  - Script uses Terraform Registry to fetch latest version
  - Script accepts and updates one provider at a time
- Providers stored in `providers.json` instead of `data.nix` to use `jq` only instead of `nix` + `jq` combination

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Fixed #98652

timstott committed:
terraform-provider.acme: 1.5.0 -> 1.5.0-patched

timstott committed:
terraform-providers: store providers as JSON

timstott committed:
terraform-providers: update providers via terraform registry

timstott commented on pkgs/applications/networking/cluster/terraform-providers/default.nix:
```diff
@@ -15,9 +15,16 @@ let
+        # Place provider in Terraform 0.13 search location
+        TF_13_PROVIDER_PATH=go/bin/${data.provider-source-address}/${version}/$(go env GOARCH)/
+        mkdir -p $TF_13_PROVIDER_PATH
+        ln -s go/bin/${repo}_v${version} $TF_13_PROVIDER_PATH
```
I went with the symlink approach to avoid making two copies of the binary, but it doesn't work.

I also question whether it's `toDrv` responsibility to make the binary available in the correct location as opposed to only building. I'm open to recommendations :upside_down_face:.

zimbatm replied:
A wrapper script that execs the providers should be a good alternative. It's the approach we are taking in https://github.com/numtide/generate-terraform-provider-shim to download providers on demand and it works fine there.

timstott commented on pkgs/applications/networking/cluster/terraform-providers/providers.json:
```diff
@@ -0,0 +1,1018 @@
+    "sha256": "0sk0pp178w03fhsb65b9mpim1l4wqfnv9r9x64kiapjnvfb1rz3j",
+    "version": "0.2.3"
+  },
+  "acme": {
+    "owner": "getstackhead",
+    "provider-source-address": "registry.terraform.io/getstackhead/acme",
+    "repo": "terraform-provider-acme",
+    "rev": "v1.5.0-patched",
+    "sha256": "1wdrjpd3l0xadsa3lqhsc9c57g8x2qkwb76q824sk8za1a7lapii",
+    "version": "1.5.0-patched"
+  },
```
I have been using the `acme` provider to test. The remaining `"provider-source-address"` will be added within this PR.

zimbatm commented on pkgs/applications/networking/cluster/terraform-providers/update-provider:
```diff
@@ -0,0 +1,55 @@
+set -euo pipefail
+
+# TODO: document usage
+provider_name="$1"
```
Do you see an easy way to list and/or update all of the providers from the registry?

timstott replied:
I'm inclined to provide two additional scripts, one to add a provider to `providers.json` from it's registry URL (as opposed to exhaustively add all providers). And one script to update all known providers.

```bash
# update-all-providers
jq 'keys[]' providers.json | xargs ./update-provider
```

I should be able to also re-use much of `update-provider` to create `add-provider registry.terraform.io/some/provider`

Any thoughts?

I will absolutely look at using `update-provider` with the `updateScript` framework :heart:

zimbatm replied:
Everything you said sounds good to me!

zimbatm reviewed (commented):
Thanks, that's great!

The question is, how complete do you want the solution to be and this depends on the energy that you have available. The current PR is already a quite good solution. It could be made better by integrating with the updateScript framework, and by allowing to get new providers from the registry.

timstott committed:
terraform: expose providers to terraform 0.13

Terraform 0.13 adopts a new filesystem layout for plugins (illustrated below).

Terraform 0.12 and earlier
`plugins-dir/terraform-provider-aws_v3.7.0`

Terraform 0.13
`plugins-dir/registry.terraform.io/hashicorp/aws/3.7.0/linux_amd64/terraform-provider-aws_v3.7.0`

To support all packaged Terraform versions a shim is created at both locations.

This approach was inspired by https://github.com/numtide/generate-terraform-provider-shim

Terraform 0.13 provider documentation https://www.terraform.io/upgrade-guides/0-13.html#new-filesystem-layout-for-local-copies-of-providers

layout terraform providers filesystem in withPlugins

timstott committed:
terraform-providers: populate provider source addresses

Added provider source address to providers hosted in the same GitHub
repository. Other providers have moved repository and would require a
patch to the package path.

timstott committed:
terraform-providers: script to update all providers

timstott commented on pkgs/applications/networking/cluster/terraform-providers/default.nix:
```diff
@@ -2,11 +2,16 @@
-  toDrv = data:
+  toDrv = name: data:
+    let
+      fallbackProviderSourceAddress = "nixpkgs/${data.owner}/${name}";
```
Provide a fallback for providers which haven't updated to use the Terraform registry. But I haven't able to make this work.
```tf
# main.tf
terraform {
  required_providers {
    acme = {
      source = "getstackhead/acme"
      version = "1.5.0-patched"
    }
    auth0 = {
      source = "nixpkgs/terraform-providers/auth0"
      version = "0.11.0"
    }
  }
}

provider "acme" {
}

provider "auth0" {
}
```

```bash
nix-shell -I nixpkgs=$PAHT_TO_NIXPKGS -p 'terraform_0_13.withPlugins (p : [p.acme p.auth0])' --command "terraform init"

Initializing the backend...

Initializing provider plugins...
- Using previously-installed getstackhead/acme v1.5.0-patched
- Finding nixpkgs/terraform-providers/auth0 versions matching "0.11.0"...

Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider
nixpkgs/terraform-providers/auth0: provider nixpkgs/terraform-providers/auth0
was not found in any of the search locations

- /nix/store/04xgzlpfc1slds41idrh5j81l8h85mpg-tf-plugin-env/bin

```

And yet
```
tree -l /nix/store/04xgzlpfc1slds41idrh5j81l8h85mpg-tf-plugin-env/bin
/nix/store/04xgzlpfc1slds41idrh5j81l8h85mpg-tf-plugin-env/bin
├── nixpkgs -> /nix/store/y16ks7jgads2y8gjzv6y7dv5sd6ydnln-terraform-provider-auth0-0.11.0/bin/nixpkgs
│   └── terraform-providers
│       └── auth0
│           └── 0.11.0
│               └── linux_amd64
│                   └── terraform-provider-auth0_v0.11.0
├── registry.terraform.io -> /nix/store/fay4phh1p79wbl4ywzf9zcvs5hcs4sf4-terraform-provider-acme-1.5.0-patched/bin/registry.terraform.io
│   └── getstackhead
│       └── acme
│           └── 1.5.0-patched
│               └── linux_amd64
│                   └── terraform-provider-acme_v1.5.0-patched
```

timstott replied:
This issue had nothing to do with the provider source address but with Terraform not following symlinks. According to the output  `only the base directory <plugin-dir> may be a symlink`, rules... are meant to be broken!

```
Initializing provider plugins...
- Finding hashicorp/cloudinit versions matching "1.0.0"...
2020/10/06 22:35:06 [WARN] Provider plugin search ignored symlink /nix/store/ws0i4lifcj0sk4x5zhxbqkyj29b5yyvp-tf-plugin-env/bin/terraform-provider-cloudinit_v1.0.0: only the base directory /nix/store/ws0i4lifcj0sk4x5zhxbqkyj29b5yyvp-tf-plugin-env/bin may be a symlink
2020/10/06 22:35:06 [WARN] Provider plugin search ignored symlink /nix/store/ws0i4lifcj0sk4x5zhxbqkyj29b5yyvp-tf-plugin-env/bin/terraform-provider-dns_v2.2.0: only the base directory /nix/store/ws0i4lifcj0sk4x5zhxbqkyj29b5yyvp-tf-plugin-env/bin may be a symlink
- Finding latest version of hashicorp/dns...

Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider
hashicorp/cloudinit: provider registry.terraform.io/hashicorp/cloudinit was
not found in any of the search locations

- /nix/store/ws0i4lifcj0sk4x5zhxbqkyj29b5yyvp-tf-plugin-env/bin


Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider hashicorp/dns:
provider registry.terraform.io/hashicorp/dns was not found in any of the
search locations

- /nix/store/ws0i4lifcj0sk4x5zhxbqkyj29b5yyvp-tf-plugin-env/bin
```

zimbatm replied:
Good investigation!

This seems to shift the implementation to the `withPlugins` function. Each plugin has to expose their attributes, and then `withPlugins` generates a trampoline script in the right location for each of the included providers.

timstott replied:
Fixed!
```tf
# main.tf
terraform {
  required_providers {
    acme = {
      source  = "getstackhead/acme"
      version = "1.5.0-patched"
    }
    auth0 = {
      source  = "nixpkgs/terraform-providers/auth0"
      version = "0.11.0"
    }
    cloudinit = {
      source  = "hashicorp/cloudinit"
      version = "1.0.0"
    }
    dns = {
      source  = "hashicorp/dns"
      version = "2.2.0"
    }
  }
}

provider "acme" {}
provider "auth0" {}
provider "cloudinit" {}
provider "dns" {}
```

```sh
nix-shell -I nixpkgs=$HOME/projects/nix/nixpkgs -p 'terraform_0_13.withPlugins (p : with p; [acme auth0 cloudinit dns])' --command "terraform init"

Initializing the backend...

Initializing provider plugins...
- Finding getstackhead/acme versions matching "1.5.0-patched"...
- Finding nixpkgs/terraform-providers/auth0 versions matching "0.11.0"...
- Finding hashicorp/cloudinit versions matching "1.0.0"...
- Finding hashicorp/dns versions matching "2.2.0"...
- Installing getstackhead/acme v1.5.0-patched...
- Installed getstackhead/acme v1.5.0-patched (unauthenticated)
- Installing nixpkgs/terraform-providers/auth0 v0.11.0...
- Installed nixpkgs/terraform-providers/auth0 v0.11.0 (unauthenticated)
- Installing hashicorp/cloudinit v1.0.0...
- Installed hashicorp/cloudinit v1.0.0 (unauthenticated)
- Installing hashicorp/dns v2.2.0...
- Installed hashicorp/dns v2.2.0 (unauthenticated)

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

![image](https://user-images.githubusercontent.com/1334474/95495381-40fd8400-0997-11eb-9b6c-9ad4f77a68d1.png)

timstott committed:
terraform: add maintainer timstott

zimbatm commented:
Result of `nixpkgs-review pr 99198` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>terraform-providers.acme</li>
    <li>terraform_plugins_test</li>
  </ul>
</details>

zimbatm commented on pkgs/applications/networking/cluster/terraform-providers/default.nix:
```diff
@@ -2,11 +2,16 @@
+  toDrv = name: data:
+    let
+      fallbackProviderSourceAddress = "nixpkgs/${data.owner}/${name}";
+      providerSourceAddress = lib.attrByPath ["provider-source-address"] fallbackProviderSourceAddress data;
```
this is a little bit cleaner, the `or` operator is triggered if the key doesn't exist.
```suggestion
      providerSourceAddress = data.provider-source-address or fallbackProviderSourceAddress;
```

timstott replied:
Much nicer :heart:. I've incorporated the changes.

zimbatm reviewed (approved):
excellent

zimbatm commented:
thanks!

jonringer commented:
specifically cd1b59476793278191c37572c8081b856512eec4 broke it

zimbatm commented:
Sorry for the inconvenience. I will have a fix ready in 3h.

voobscout commented:
Please look into it again.
shell.nix:
```nix
let
  pkgs = import (builtins.fetchTarball "https://api.github.com/repos/nixos/nixpkgs/tarball/master") { };
in
pkgs.mkShell {
  buildInputs = [
    (pkgs.terraform_0_13.withPlugins (p: [
      p.null
      p.helm
      p.vultr
      p.digitalocean
      p.kubernetes
    ]))
  ];
}
```
```bash
nix-shell --run "terraform init"
```
produces:
```
Initializing the backend...

Initializing provider plugins...
- Using previously-installed hashicorp/null v2.1.2
- Using previously-installed hashicorp/helm v1.2.2
- Finding latest version of hashicorp/kubernetes...
- Finding latest version of hashicorp/digitalocean...
- Finding latest version of hashicorp/vultr...

Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider
hashicorp/kubernetes: provider registry.terraform.io/hashicorp/kubernetes was
not found in any of the search locations

- /nix/store/6bgi4zxw3xvpni361hs8l7xf9d33cy3d-terraform-0.13.5/plugins


Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider
hashicorp/digitalocean: provider registry.terraform.io/hashicorp/digitalocean
was not found in any of the search locations

- /nix/store/6bgi4zxw3xvpni361hs8l7xf9d33cy3d-terraform-0.13.5/plugins


Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider
hashicorp/vultr: provider registry.terraform.io/hashicorp/vultr was not found
in any of the search locations

- /nix/store/6bgi4zxw3xvpni361hs8l7xf9d33cy3d-terraform-0.13.5/plugins
```

looking at *plugins/registry.terraform.io*, we find 2 dirs - **hashicorp** and **nixpkgs**.
```bash
ls /nix/store/6bgi4zxw3xvpni361hs8l7xf9d33cy3d-terraform-0.13.5/plugins/registry.terraform.io/
# dr-xr-xr-x 3 root 3 Jan  1  1970 helm/
# dr-xr-xr-x 3 root 3 Jan  1  1970 null/
```
Not found plugins ended up in *plugins/registry.terraform.io/nixpkgs*:
```bash
ls /nix/store/6bgi4zxw3xvpni361hs8l7xf9d33cy3d-terraform-0.13.5/plugins/registry.terraform.io/nixpkgs
# dr-xr-xr-x 3 root 3 Jan  1  1970 digitalocean/
# dr-xr-xr-x 3 root 3 Jan  1  1970 kubernetes/
# dr-xr-xr-x 3 root 3 Jan  1  1970 vultr/
```

arcnmx commented:
Currently you need something like...
```
terraform.required_providers = mapAttrs (name: plugin: {
  version = plugin.version;
  source = plugin.provider-source-address or "nixpkgs/${name};
}) pkgs.terraform.plugins;
```

for providers without the source address set, or just:

```
terraform {
  required_providers {
    helm = {
      source = "nixpkgs/helm"
    }
  }
}
```

voobscout commented:
attribute 'required_providers' missing, can you please supply a full example?

timstott commented:
@voobscout the snippet below, to execute in the `nix repl`, will output the provider definition in `nix` and will have to be adapted to `HCL` in your `main.tf`

```
$ nix repl
:l https://api.github.com/repos/nixos/nixpkgs/tarball/master

providers = (lib.getAttrs
  ["null" "helm" "vultr" "digitalocean" "kubernetes"]
  pkgs.terraform-providers)

required_providers = lib.mapAttrs (name: plugin: {
  version = plugin.version;
  source = plugin.provider-source-address or "nixpkgs/${name}";
}) providers

:p { terraform = { inherit required_providers; }; }
```

This is highlighting an improvement I have been meaning to make, to include the provider declaration snippet in the package description (much like on the Terraform registry - screenshot below).

![image](https://user-images.githubusercontent.com/1334474/97978467-814cf800-1dc5-11eb-9099-618624cf376e.png)

voobscout commented:
That makes a lot of sense now, thank you @timstott!

sevenfourk commented:
guys, maybe you can point me what i'm missing here: 

```console
- Using previously-installed nixpkgs/kubernetes v1.11.3
- Using previously-installed hashicorp/local v2.0.0
- Finding hashicorp/kubernetes versions matching "~> 1.10, >= 1.11.1"...

Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider
hashicorp/kubernetes: provider registry.terraform.io/hashicorp/kubernetes was
not found in any of the search locations

- /nix/store/cfldgk8hx5zfc3nzij52h149q6l1bnns-terraform-0.13.5/plugins

[terragrunt] 2020/11/22 18:25:40 Hit multiple errors:
exit status 1

❯ cat versions.tf
terraform {
  required_providers {
    helm = {
      source = "nixpkgs/kubernetes"
    }
  }
}
````

sorry for posting in merged pull request.

arcnmx commented:
> guys, maybe you can point me what i'm missing here:

```
    helm = {
      source = "nixpkgs/kubernetes"
    }
```
I'm pretty sure you want that to be `kubernetes = {`

sevenfourk commented:
sorry to confuse, I had correct versions.tf, was just copy pasting from your comment before posting my comment:

```console
- Finding hashicorp/kubernetes versions matching ">= 1.11.1"...

Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider
hashicorp/kubernetes: provider registry.terraform.io/hashicorp/kubernetes was
not found in any of the search locations

- /nix/store/cfldgk8hx5zfc3nzij52h149q6l1bnns-terraform-0.13.5/plugins

[terragrunt] 2020/11/22 19:58:44 Hit multiple errors:
exit status 1

❯ cat versions.tf
terraform {
  required_providers {
    kubernetes = {
      source = "nixpkgs/kubernetes"
    }
  }
}
```

sevenfourk commented:
so i did edit also eks module:

```console
❯ cat .terraform/modules/eks/versions.tf
terraform {
  required_providers {
    kubernetes = {
      source = "nixpkgs/kubernetes"
    }
  }
}
```

and did this trick:

```console
terragrunt state replace-provider registry.terraform.io/hashicorp/kubernetes nixpkgs/kubernetes
```

which is incredibly _ugly_.

grahamc commented:
I had a pretty hard time with this, and want to call out the replace-provider of the previous comment.

I had this in my `.tf` files:

```
terraform {
  required_providers {
    aws = {
      source = "hashicorp/aws"
    }

    local = {
      source = "hashicorp/local"
    }

    null = {
      source = "hashicorp/null"
    }
  }
}
```

I'd then run `terraform init` and it'd look fine:

```
$ terraform init ./terraform
Initializing modules...
- xxx in xxx/terraform/xxx

Initializing the backend...

Initializing provider plugins...
- Finding latest version of hashicorp/local...
- Finding latest version of hashicorp/null...
- Finding latest version of hashicorp/aws...
- Installing hashicorp/local v1.4.0...
- Installed hashicorp/local v1.4.0 (unauthenticated)
- Installing hashicorp/null v2.1.2...
- Installed hashicorp/null v2.1.2 (unauthenticated)
- Installing hashicorp/aws v2.65.0...
- Installed hashicorp/aws v2.65.0 (unauthenticated)

The following providers do not have any version constraints in configuration,
so the latest version was installed.

To prevent automatic upgrades to new major versions that may contain breaking
changes, we recommend adding version constraints in a required_providers block
in your configuration, with the constraint strings suggested below.

* hashicorp/aws: version = "~> 2.65.0"
* hashicorp/local: version = "~> 1.4.0"
* hashicorp/null: version = "~> 2.1.2"

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

and then I'd go to do an apply, and it fails:

```
$ terraform apply \
	-state="./xxx/terraform.tfstate" \
	-input=false \
	-var-file="./xxx/terraform.tfvars.json" \
	./terraform

Error: Could not load plugin


Plugin reinitialization required. Please run "terraform init".

Plugins are external binaries that Terraform uses to access and manipulate
resources. The configuration provided requires plugins which can't be located,
don't satisfy the version constraints, or are otherwise incompatible.

Terraform automatically discovers provider requirements from your
configuration, including providers used in child modules. To see the
requirements and constraints, run "terraform providers".

2 problems:

- Failed to instantiate provider "registry.terraform.io/-/aws" to obtain
schema: unknown provider "registry.terraform.io/-/aws"
- Failed to instantiate provider "registry.terraform.io/-/local" to obtain
schema: unknown provider "registry.terraform.io/-/local"
```


My shell.nix looked like this:

```
{ pkgs ? import <nixpkgs> { } }:
pkgs.mkShell {
  buildInputs = [
    (pkgs.terraform_0_13.withPlugins
      (p: [ p.aws p.template p.external p.null p.local ]))
  ];
}
```

The issue is actually a problem with the 0.12 to 0.13 migration, NOT a problem with this PR.

Our providers were considered "in-house" before, and have less information than terraform wants for the automatic upgrade. This means we need to follow this: https://www.terraform.io/upgrade-guides/0-13.html#in-house-providers

I needed to run:

```
$ terraform state replace-provider -state="./xxx/terraform.tfstate" 'registry.terraform.io/-/aws' 'hashicorp/aws'

$ terraform state replace-provider -state="./xxx/terraform.tfstate" 'registry.terraform.io/-/local' 'hashicorp/local'
```


and that fixed the issue with:

```
- Failed to instantiate provider "registry.terraform.io/-/aws" to obtain
schema: unknown provider "registry.terraform.io/-/aws"
```

grahamc commented:
Note if you see an error like:

```
$ terraform init

Initializing the backend...

Initializing provider plugins...
- Finding latest version of hashicorp/aws...

Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider hashicorp/aws:
provider registry.terraform.io/hashicorp/aws was not found in any of the
search locations

- /nix/store/fam6azhh0wlpaxa27rbxkjyqzrla50xn-tf-plugin-env/bin
```

it might be because you're on 20.09 and 20.09 doesn't have this PR applied.

ivankovnatsky commented:
sorry to bring it here, if you have something to say from your experience please share your thoughts here:  https://discourse.nixos.org/t/public-cloud-infrastructure-people-whats-you-workflow-with-terraform-terragrunt-providers/11669/4

for a couple of months trying to use nixos provided terraform and providers i faced multiple issues contributing to many people infra projects, like:
* figuring out how to configure tf in nixos to install only needed providers and not to install all of them
* issues mentioned in this pr
* figuring out how to pin the tf version (i tried to overlay 0.14.4 and it failed to build, i did not look into it too deeply)
* nixos built providers, which might be not always up to date

also tf 0.14 introduced a lock file to keep providers hashes consistent for all contributors in repo, in case of nixos providers these hashes would be different, meaning we can't use this lock file.

again, i might using it all wrong, please share your workflow if my attitude towards nixos tf and providers are rather wrong.

and yeah, just wondering what is the use case to manage/build providers in nixos, rather than using hashicorp ones? (that might be due to nixos source model, if that's the case I could not agree more).

but for me, it brings much more inconvenience (maybe due to my workflow).

anandsuresh commented:
I'm seeing this in nixpkgs v21.05. In my case, I'm building my TF modules from source, and they are being dropped into the `/nix/store/qz2ahhd4g3nsrjlykwgs54ssb42lz8ck-terraform-0.15.4/plugins`. I can see the compiled providers there, but I am still receiving an error.

```
$ terraform  init

Initializing the backend...

Initializing provider plugins...
- Finding cloudflare/cloudflare versions matching "2.23.0"...
╷
│ Error: Failed to query available provider packages
│
│ Could not retrieve the list of available versions for provider cloudflare/cloudflare: provider registry.terraform.io/cloudflare/cloudflare was not found in any of
│ the search locations
│
│   - /nix/store/qz2ahhd4g3nsrjlykwgs54ssb42lz8ck-terraform-0.15.4/plugins
╵
```

The suggested fix to run `terraform state replace-provider` does not apply to my case since this is a new codebase and there is no `.tfstate` file at this stage. Has anyone else encountered this issue before?

1000101 commented:
@anandsuresh Have you been able to fix the issue? I've bumped into just now and defining `terraform required_providers` in my `config.tf` did the trick for me, e.g.:

```
terraform {
    required_providers {
      cloudflare {
        source = "cloudflare/cloudflare"
      }
    }
  }
```
